dtyper.py

```diff
@@ -1,23 +1,39 @@
-from dataclasses import make_dataclass, field
-from functools import wraps
+from __future__ import annotations
+
 import inspect
+from dataclasses import field, make_dataclass
+from functools import wraps
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from typing import Callable, Optional, Type, TypeVar, Union
+
+    from typing_extensions import ParamSpec
+
+    P = ParamSpec("P")
+    R = TypeVar("R")
 
 __version__ = '0.9.0'
 
 
 @wraps(make_dataclass)
-def dataclass(typer_command, base=None, **kwargs):
-    def to_field(k, v, default):
-        return (k, v) if default is ... else (k, v, field(default=default))
+def dataclass(
+    typer_command: Callable[P, R], base: Optional[Type] = None, **kwargs
+) -> Callable[[Union[Type, Callable]], Type]:
 
     if base is not None:
         kwargs['bases'] = *kwargs.get('bases', ()), base
 
     kwargs.setdefault('cls_name', typer_command.__name__)
-    kwargs['fields'] = [to_field(*i) for i in _params(typer_command)]
+    kwargs['fields'] = [
```

Comment:
This is a pessimization of the code from before in terms of being readable.  For example, note that the if/else then are in the reverse order of the standard reading!

The reason for `to_field` is to make it easy to read. Efficiency is of zero interest since this is called one time per program/function.

I suggest something like

    sig = _fixed_signature(typer_command)
    kwargs['fields'] = [to_field(*i) for i in sig.parameters.values()]

and put all that logic into `to_field`...